Enhancement/updated spec (#11)
gardener/machine-spec: spec.md

> # Container Machine Interface (CMI)

Author: Any suggestions for a better name are welcome here. Preferably something along the lines of CMI.

Reviewer: How about CMI but *Common* Machine Interface? 😄

Reviewer 2: `Container Machine Interface` is perfect IMO.

Author: > Container Machine Interface is perfect IMO.

But @stoyanr this has very little to do with containers right?

Authors:

* Prashanth <<prashanth@sap.com>> (@prashanth26)
* Hardik Dodiya <<hardik.dodiya@sap.com>> (@hardikdr)

> * Gaurav Gupta <<gaurav.gupta07@sap.com>> (@ggaurav@10)

Author: ```suggestion
* Gaurav Gupta <<gaurav.gupta07@sap.com>> (@ggaurav10)
```

* It is also inspired from an historic internal documentation provided by Stoyan Rachev <<s.rachev@sap.com>> (@stoyanr).

## Table of contents

TODO

## Notational Conventions

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" are to be interpreted as described in [RFC 2119](http://tools.ietf.org/html/rfc2119) (Bradner, S., "Key words for use in RFCs to Indicate Requirement Levels", BCP 14, RFC 2119, March 1997).

The key words "unspecified", "undefined", and "implementation-defined" are to be interpreted as described in the [rationale for the C99 standard](http://www.open-std.org/jtc1/sc22/wg14/www/C99RationaleV5.10.pdf#page=18).

An implementation is not compliant if it fails to satisfy one or more of the MUST, REQUIRED, or SHALL requirements for the protocols it implements.
An implementation is compliant if it satisfies all the MUST, REQUIRED, and SHALL requirements for the protocols it implements.

## Terminology

| Term            | Definition                                       |
|-----------------|--------------------------------------------------|

> | CR              | Custom Resource (CR) defined by an cluster admin using the Kubernetes CR primitive. |

Reviewer 3: ```suggestion
| CR              | Custom Resource (CR) is defined by a cluster admin using the Kubernetes Custom Resource Definition  primitive. |
```

> | VM              | An Virtual Machine (VM) provisioned and managed by a provider. It could also refer to a physical machine in case of an bare metal provider.|

Reviewer 3: ```suggestion
| VM              | A Virtual Machine (VM) provisioned and managed by a provider. It could also refer to a physical machine in case of an bare metal provider.|

> | VM              | A Virtual Machine (VM) provisioned and managed by a provider. It could also refer to a physical machine in case of an bare metal provider.|

Reviewer 4: s/an bare/a bare/

Author: Done.

| gRPC            | [gRPC Remote Procedure Calls](https://grpc.io/) |

> | MCM             | [Machine Controller Manager (MCM)](https://github.com/gardener/machine-controller-manager) is the controller used to manage VMs as an Custom Resource (CR) in Kubernetes. The MCM is made up of two containers - `CMI-Client` and `CMI-Plugin`. |

Reviewer 3: ```suggestion
| MCM             | [Machine Controller Manager (MCM)](https://github.com/gardener/machine-controller-manager) is the controller used to manage VMs as a Custom Resource (CR) in Kubernetes. The MCM is made up of two containers - `CMI-Client` and `CMI-Plugin`. |
```

> | Machine         | Machine refers to a VM that is provisioned/managed by the MCM. It typically is the metadata used to store/represent an Virtual Machine |

Reviewer 3: ```suggestion
| Machine         | Machine refers to a VM that is provisioned/managed by MCM. It typically describes the metadata used to store/represent an Virtual Machine |
```

> | Machine         | Machine refers to a VM that is provisioned/managed by MCM. It typically describes the metadata used to store/represent an Virtual Machine |

Reviewer 4: s/an Virtual/a Virtual/

Author: Done.

| Node            | Native kubernetes `Node` object. The objects you get to see when you do a "kubectl get nodes". Although nodes can be either physical/virtual machines, for the purposes of our discussions it refers to a VM. |
| CMI-client      | CMI-Client contains the *core generic logic used to manage machines*. It doesn't contain any cloud specific code to manage VMs. It delegates the job of creating/updating/deleting VMs to the `CMI-Plugin` using gRPC primitives. |

> | CMI-Plugin      | CMI-Plugin (or) Plugin (or) CMI-Server is the plugin responsible for accepting gRPC calls from CMI-Client and invoking provider specific functionality to do the same. An simple example could be creation/deletion of VM on the provider. |

Reviewer 3: ```suggestion
| CMI-Plugin      | CMI-Plugin (or) Plugin (or) CMI-Server is the plugin responsible for accepting gRPC calls from CMI-Client and invoking provider specific functionality to do the same. A simple example could be creation/deletion of VM on the provider. |
```

## Solution Overview

This specification defines an interface along with the minimum operational and packaging recommendations for a VM provider to implement a CMI compatible plugin.
The interface declares the RPCs that a plugin MUST expose: this is the **primary focus** of the CMI specification.

### Architecture

The primary focus of this specification is on the **protocol** between a CMI-Client and a Plugin.
It SHOULD be possible to ship CMI-Client compatible Plugins for a variety of deployment architectures.
A CMI-Client SHOULD be equipped to handle both centralized and headless plugins, as well as split-component and unified plugins.
Several of these possibilities are illustrated in the following figures.

```
                             MCM Architecture
+-------------------------------------------+
|                                           |
|  +------------+           +------------+  |
|  |   CMI      |   gRPC    |   CMI      |  |
|  |   Client   +----------->   Plugin   |  |
|  +------------+           +------------+  |
|                                           |
+-------------------------------------------+

Figure 1: CMI client and CMI Plugin together make up

> the MCM. Both of them are typically expected to run

Reviewer: ```suggestion
the machine part of MCM. Both of them are typically expected to run
```

> as two containers is a same pod. However, both might

Reviewer 4: s/is/in/

Author: Done.

run in different pods as well.
```

### Machine Lifecycle

> ```

Reviewer 4: IMHO, following should be the lifecycle of a machine as far as CMI client is concerned. This clearly shows the interfaces that MCM uses to manage lifecycle of a machine.

```
       +-+                      +-+
       |X|   +----------------->| |
       +++   |                  +++
        |    |                   ^
        |    |                   | Delete
        |    |                   | Machine
        |    |                   |
 Create |    | Delete       +----+-----+
Machine |    | Machine      |  STOPPED |
        |    |              +----^-----+
        |    |                   |
        |    |                   | Shutdown
    +---v----+---+               | Machine
    |   CREATED  |---------------+
    +------------+
```

Author: Done.

  CreateMachine +------------+ DeleteMachine
 +------------->|  PENDING   +--------------+
 |              +---+----+---+              |
 |               VM |    | VM               v
+++              is |    | is              +++
|X|              Up |    | Down            | |
+-+             +---v----+---+             +-+
                | AVAILABLE  |
                +---+----^---+
                 VM |    | VM
        is ready to |    | is not ready to
      schedule pods |    | schedule pods
                +---v----+---+
                |   READY    |
                +------------+

Figure 2: The lifecycle of a dynamically provisioned machine, from
creation to destruction.
```

The above diagrams illustrate a general expectation with respect to how a MCM manages the lifecycle of a machine via the API presented in this specification.

> Plugins SHOULD expose all RPCs for an interface: Controller plugins SHOULD implement all RPCs for the `Machine` service.

Reviewer 3: As it is stated as a 'SHOULD' requirement and not a 'SHALL' requirement, will it be possible that plugin can lead a machine to an orphan state if it is allowed to implement some and not all of the RPCs. 
For e.g. if it implements CreateMachine but not DeleteMachine or what if it implements DeleteMachine but not ShutdownMachine or take a case where it doesn't implement GetListOfVolumeIDsForExistingPVs.

Maybe this will come out during test as well, but should be made explicit that just returning unsupported RPCs with CALL_NOT_IMPLEMENTED is not sufficient for compliance, the CMI-Plugin provider should ensure the lifecycle management of the machines with validated state transitions as described in the diagram above.

WDYT ?

Author: Yes, I agree with you. We need a harder restriction on the required methods and optional ones. Let me try to adapt the proposal to the same.

Reviewer 2: I think should be a minimum set of capabilities that are mandatory since without them, the CMI client couldn't do a meaningful job. I think this should include `CreateMachine` and `DeleteMachine`, not sure about `GetMachine`. The rest should be optional.

Author: ```suggestion
Plugins SHOULD expose a minimum set of a REQUIRED RPCs (`CreateMachine, `DeleteMachine`). Other RPCs (like `GetMachine`, `ShutDownMachine`, `ListMachines` and `GetVolumeIDs`) CAN be optionally implemented. However, it is highly recommended to implement these OPTIONAL RPCs for an optimized working of CMI Client. For more details on the optimizations provided for each OPTIONAL RPC, please refer the sections for each of the RPC descriptions. 
```

How about something like this?

Each VM provider MUST provide two sets of RPCs:

* **Identity Service**: These are the set of identity RPCs to be implemented by the CMI-PLugin.
* **Machine Service**: The CMI-Plugin implement this sets of machine RPCs.

```protobuf

service Identity {
    rpc GetPluginInfo(GetPluginInfoRequest)
        returns (GetPluginInfoResponse) {}

    rpc GetPluginCapabilities(GetPluginCapabilitiesRequest)
        returns (GetPluginCapabilitiesResponse) {}

    rpc Probe (ProbeRequest)
        returns (ProbeResponse) {}
}

service Machine {
    rpc CreateMachine (CreateMachineRequest)
        returns (CreateMachineResponse) {}

    rpc DeleteMachine (DeleteMachineRequest)
        returns (DeleteMachineResponse) {}

    rpc GetMachine (GetMachineRequest)
        returns (GetMachineResponse) {}

    rpc ListMachines (ListMachinesRequest)
        returns (ListMachinesResponse) {}

    rpc ShutDownMachine (ShutDownMachineRequest)
        returns (ShutDownMachineResponse) {}

    rpc GetListOfVolumeIDsForExistingPVs(GetListOfVolumeIDsForExistingPVsRequest)
        returns (GetListOfVolumeIDsForExistingPVsResponse) {}
}

```

#### Concurrency

> In general the CMI-Client is responsible for ensuring that there is no more than one call “in-flight” per machine at a given time.

Reviewer 4: This line is taken from CSI as is. It may be applicable to CSI, however, are we sure that MCM adheres to this? Could safety controller and/or machine controller (or any other controller loop inside MCM) make simultaneous calls (eg. GetMachine) to the plugin. Even if not, do we need this kind of behaviour from the plugins? Eg. - simultaneous GetMachine calls for the same machine seem harmless.

Reviewer 2: I think all of this is needed. At least in our GMP-related implementation it's a very legitimate concern. The CMI plugin should ensure that normally it doesn't make simultaneous calls. However it may "lose state", e.g. crash and be restarted, and issue a parallel call for this reason, so the plugin should in principle be prepared to handle the situation as described. Instead of "call" you may wish to talk about "operation" and include only those operations that change the machine state.

> However, in some circumstances, the CMI-Client maybe lose state (for example when the CMI-Client crashes and restarts), and MAY issue multiple calls simultaneously for the same machine.

Reviewer 4: s/maybe/may/

Author: Done.

The general size limit for a particular field MAY be overridden by specifying a different size limit in said field's description.
Unless otherwise specified, fields SHALL NOT exceed the limits documented here.
These limits apply for messages generated by both the CMI-Client and CMI-plugins.

| Size       | Field Type          |
|------------|---------------------|
| 128 bytes  | string              |
| 4 KiB      | map<string, string> |

##### `REQUIRED` vs. `OPTIONAL`

* A field noted as `REQUIRED` MUST be specified, subject to any per-RPC caveats; caveats SHOULD be rare.
* A `repeated` or `map` field listed as `REQUIRED` MUST contain at least 1 element.
* A field noted as `OPTIONAL` MAY be specified and the specification SHALL clearly define expected behavior for the default, zero-value of such fields.

Scalar fields, even REQUIRED ones, will be defaulted if not specified and any field set to the default value will not be serialized over the wire as per [proto3](https://developers.google.com/protocol-buffers/docs/proto3#default).

#### Timeouts

Any of the RPCs defined in this spec MAY timeout and MAY be retried.

> The CMI-Client may choose the maximum time it is willing to wait for a call, how long it waits between retries, and how many time it retries (these values are not negotiated between plugin and CO).

Reviewer: > between plugin and CO

What's CO?

Author: ```suggestion
The CMI-Client may choose the maximum time it is willing to wait for a call, how long it waits between retries, and how many time it retries (these values are not negotiated between plugin and CMI-Client).
```

> The CMI-Client may choose the maximum time it is willing to wait for a call, how long it waits between retries, and how many time it retries (these values are not negotiated between plugin and CO).

Reviewer 3: Will it be better to have timeout and retry configurable per provider, for e.g. DeleteMachine might take more time  for a particular provider, as in principle it is not in the purview of cmi-client to be aware of any provider specifics but do you see it might make it more optimal and easier to have these semantics returned as part of GetPluginInfoResponse?

Reviewer 2: @ashwani2k I don't think this is needed, as far as MCM can be configured differently for the different providers, which is typically the case in a Gardener setup.

Idempotency requirements ensure that a retried call with the same fields continues where it left off when retried.

### Error Scheme

All CMI API calls defined in this spec MUST return a [standard gRPC status](https://github.com/grpc/grpc/blob/master/src/proto/grpc/status/status.proto).
Most gRPC libraries provide helper methods to set and read the status fields.

The status `code` MUST contain a [canonical error code](https://github.com/grpc/grpc-go/blob/master/codes/codes.go). CMI-Clients MUST handle all valid error codes. Each RPC defines a set of gRPC error codes that MUST be returned by the plugin when specified conditions are encountered. In addition to those, if the conditions defined below are encountered, the plugin MUST return the associated gRPC error code.

| Condition | gRPC Code | Description | Recovery Behavior |
|-----------|-----------|-------------|-------------------|
| Missing required field | 3 INVALID_ARGUMENT | Indicates that a required field is missing from the request. More human-readable information MAY be provided in the `status.message` field. | Caller MUST fix the request by adding the missing required field before retrying. |

> | Invalid or unsupported field in the request | 3 INVALID_ARGUMENT | Indicates that the one or more fields in this field is either not allowed by the Plugin or has an invalid value. More human-readable information MAY be provided in the gRPC `status.message` field. | Caller MUST fix the field before retrying. |

Reviewer 3: ```suggestion
| Invalid or unsupported field in the request | 3 INVALID_ARGUMENT | Indicates that the one or more fields in this field is either not allowed by the Plugin or has an invalid value. More human-readable information MAY be provided in the gRPC `status.message` field. | Caller MUST either replace the field with a valid or supported field or remove the invalid or unsupported field before retrying. |
```

| Permission denied | 7 PERMISSION_DENIED | The Plugin is able to derive or otherwise infer an identity from the secrets present within an RPC, but that identity does not have permission to invoke the RPC. | System administrator SHOULD ensure that requisite permissions are granted, after which point the caller MAY retry the attempted RPC. |

> | Operation pending for machine | 10 ABORTED | Indicates that there is already an operation pending for the specified machine. In general the CMI-Client is responsible for ensuring that there is no more than one call "in-flight" per machine at a given time. However, in some circumstances, the CMI-Client MAY lose state (for example when the CMI-Client crashes and restarts), and MAY issue multiple calls simultaneously for the same machine. The Plugin, SHOULD handle this as gracefully as possible, and MAY return this error code to reject secondary calls. | Caller SHOULD ensure that there are no other calls pending for the specified machine, and then retry with exponential back off. |

Reviewer: > Caller SHOULD ensure that there are no other calls pending for the specified machine, and then retry with exponential back off.

We do not implement the exponential back off, do we? Should we remove that part?

Author: I was wondering if we could enforce exponential backoff in MCM for errored retries of machine objects. This would avoid queuing machines too often. Especially in cases like machines failing to join due to network issues, causing it to recreate several machines.

Reviewer: In that case, shall we split the sentence and keep the SHOULD only for the ensuring no other calls are pending part and say that it is *recommended* that the retry should be governed by an exponential backoff?

Author: Yes sure. Makes sense. Will create a commit with all your suggested changes soon.

Reviewer 2: I also think it's important to keep the handling of `ABORTED` as this situation can actually occur, and even explicitly describe it in the tables listing the errors of different calls, see below. Whether the backoff is exponential or not is a detail and should be omitted IMO. Ideally timeout, number of retries, or even retry strategy (exponential or not) should be configurable on MCM side.

Author: @stoyanr - Apologies for the delay in response. I think your idea of having configurable retry strategies on the MCM makes sense. Will Remove the details about retries in this section. Will try to describe the retry strategy in another section. 

WDYT @amshuman-kr , @stoyanr ?

Author: Have made the changes - https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR225. And have added a section on retry - https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR205-R212.

| Call not implemented | 12 UNIMPLEMENTED | The invoked RPC is not implemented by the Plugin or disabled in the Plugin's current mode of operation. | Caller MUST NOT retry. Caller MAY call `GetPluginCapabilities` to discover Plugin capabilities in future |

> | Not authenticated | 16 UNAUTHENTICATED | The invoked RPC does not carry secrets that are valid for authentication. | Caller SHALL either fix the secrets passed to the RPC, or otherwise regalvanize said secrets such that they will pass authentication by the Plugin for the attempted RPC, after which point the caller MAY retry the attempted RPC. |

Reviewer 3: `The invoked RPC does not carry secrets that are valid for authentication`
Can the plugin provider send any sensitive data as part of status.message response?  Just curious what is the propagation path of status.message and what human readable contents can it contain. May be I'm over-engineering but just thinking if it should be made explicit for these calls.  Although we are explicit cmi-client will not log any secret related information but we might have standard handling of status.message which might leak these.

Author: The plugin should definitely not leak secrets/credentials through `logs` or `status.message`. I will try add a word about the `status.message` to not pass sensitive data as well.

Author: I have added the following in a followup PR.
```
Make sure to not pass any sensitive data through plain text messages. Also make sure to not log any such sensitive data.
```

### Secrets Requirements

Secrets MAY be required by plugin to complete a RPC request.
A secret is a string map where the key identifies the name of the secret (e.g. "username" or "password"), and the value contains the secret data (e.g. "bob" or "abc123"). These `key-value` pairs are used to authenticate the request at the provider.
Each key MUST consist of alphanumeric characters, '-', '_' or '.'.
Each value MUST contain a valid string.
An VM provider MAY choose to accept binary (non-string) data by using a binary-to-text encoding scheme, like base64.
An VM provider SHALL advertise the requirements for required secret keys and values in documentation.
CMI-Client SHALL permit passing through the required secrets.
This information is sensitive and MUST be treated as such (not logged, etc.) by the CMI-Client.

### MachineClass Resources

MCM introduces the CRD `MachineClass`. This is a blueprint for creating machines that join a certain cluster as nodes in a certain role. The plugin only works with `MachineClass` resources that have the structure described here.

#### ProviderSpec

> The `MachineClass` resource contains a `providerSpec` field that is passed in the `ProviderSpec` request field to CMI methods such as [CreateMachine](#createmachine). The `ProviderSpec` can be though of a machine template to which the VM specification must be adopted from. It can contain key-value pairs of these specs. An example for these key-value pairs are given below.

Reviewer: ```suggestion
The `MachineClass` resource contains a `providerSpec` field that is passed in the `ProviderSpec` request field to CMI methods such as [CreateMachine](#createmachine). The `ProviderSpec` can be thought of as a machine template from which the VM specification must be adopted. It can contain key-value pairs of these specs. An example for these key-value pairs are given below.
```

| Parameter | Mandatory | Type | Description |
|---|---|---|---|
| `vmPool` | Yes | `string` | VM pool name, e.g. `TEST-WOKER-POOL` |
| `size` | Yes | `string` | VM size, e.g. `xsmall`, `small`, etc. Each size maps to a number of CPUs and memory size. |
| `rootFsSize` | No | `int` | Root (`/`) filesystem size in GB |
| `tags` | Yes | `map` | Tags to be put on the created VM |

Most of the `ProviderSpec` fields are not mandatory. If not specified, the plugin passes an empty value in the respective `Create VM` parameter.

> The `tags` can be used to identify machines by ID

Reviewer: This is a bit confusing. What is meant by ID? If it is machine ID then the statement need not be true, right?

Author: ```suggestion
The `tags` can be used to map a VM to its corresponding machine object's name
```
Does this sound better?

- `cloud-init` scripts used to initialize a new VM. The cloud-init script is expected to contain scripts to initialize the Kubelet and make it join the cluster.

#### Identifying Cluster Machines

To implement certain methods, the plugin should be able to identify all machines associated with a particular Kubernetes cluster. This can be achieved using one/more of the below mentioned ways:

* Names of VMs created by the plugin are prefixed by the cluster ID specified in the [ProviderSpec](#providerspec).
* VMs created by the plugin are tagged with the special tags like `kubernetes.io/cluster` (for the cluster ID) and `kubernetes.io/role` (for the role), specified in the [ProviderSpec](#providerspec).
* Mapping `Resource Groups` to individual cluster.

### Identity Service RPC

Identity service RPCs allow a CMI-Client to query a CMI-Plugin for capabilities, health, and other metadata.
The general flow of the success case MAY be as follows (protos illustrated in YAML for brevity):

1. CMI-Client queries metadata via Identity RPC.

```
   # CMI-Client --(GetPluginInfo)--> Plugin
   request:
   response:
      name: org.foo.whizbang.super-plugin

>       vendor_version: blue-green

Reviewer 4: s/vendor_version/version/

Author: Done.

      manifest:
        baz: qaz
```

2. CMI-Client queries available capabilities of the plugin.

```
   # CMI-Client --(GetPluginCapabilities)--> Plugin
   request:
   response:
     capabilities:
       - service:
           type: CONTROLLER_SERVICE

> ```

Reviewer 4: Response should be:
```
   response:
     capabilities:
       - rpc:
           type: CREATE_MACHINE
```

Author: Done.

3. CMI-Client queries the readiness of the plugin.

```
   # CMI-Client --(Probe)--> Plugin
   request:
   response: {}
```

#### `GetPluginInfo`

```protobuf
message GetPluginInfoRequest {
    // Intentionally empty.
}

message GetPluginInfoResponse {
    // The name MUST follow domain name notation format
    // (https://tools.ietf.org/html/rfc1035#section-2.3.1). It SHOULD
    // include the plugin's host company name and the plugin name,
    // to minimize the possibility of collisions. It MUST be 63
    // characters or less, beginning and ending with an alphanumeric
    // character ([a-z0-9A-Z]) with dashes (-), dots (.), and
    // alphanumerics between.
    // This field is REQUIRED.
    string name = 1;

    // The version specifies the plugin version
    // Value of this field is opaque to the CMI-Client.
    // This field is REQUIRED.
    string version = 2;

    // manifest contains a map of key-value pairs to pass
    // any additonal information about the plugin.
    // Values are opaque to the CMI-Client.
    // This field is OPTIONAL.

>     map<string, string> manifest = 3;

Reviewer 3: Can we explain the usage of manifest as it is opaque to the CMI-CLIENT, what kind of data is returned in the manifest.

Author: The idea is for the manifest to be a map of key-value pairs. This will contain any addtional information on the plugin. E.g.: key = start-time, value = timestamp (xx.xx.xxxx xx:xx:xx IST).

Author: Have added an example for the usage.

}
```

##### GetPluginInfo Errors

If the plugin is unable to complete the GetPluginInfo call successfully, it MUST return a non-ok gRPC code in the gRPC status.

#### `GetPluginCapabilities`

> This REQUIRED RPC allows the CMI-Client to query the supported capabilities of the Plugin "as a whole": it is the grand sum of all capabilities.

Reviewer 3: This is inline with my earlier comment, is this get call used just to get what calls are supported by plugin or will the client also check on the returned capabilities if the plugin is semantically compliant as well.

Author: I will try to draft this into more concrete sentences with stronger requirements and constraints.

Author: How about this -https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR122

All CMI-Plugins of the same version (see `version` of `GetPluginInfoResponse`) of the Plugin SHALL return the same set of capabilities.

```protobuf

message GetPluginCapabilitiesRequest {
    // Intentionally empty.
}

message GetPluginCapabilitiesResponse {

>     // All the capabilities that the controller service supports.

Reviewer 4: s/controller/machine/

Author: Done.

    // This field is OPTIONAL.
    repeated PluginCapability capabilities = 1;
}

// Specifies a capability of the CMI-Plugin.
message PluginCapability {
    message RPC {
        enum Type {

>             // UNKNOWN is used to specific an capaability beyond the set

Reviewer 4: s/capaability/capability/

Author: Done.

            // provided below
            UNKNOWN = 0;

            // CREATE_MACHINE tells that the plugin implements the
            // CreateMachine() interface

>             CREATE_MACHINE = 1;

Reviewer 2: As mentioned above, I would remove `CREATE_MACHINE` and `DELETE_MACHINE` and treat them as core capabilities that **must** be implemented. Also, I don't see a capability for `LIST_MACHINES` - is this intentional? I believe this call should be treated as non-core and therefore a capability for it should exist (but I may be wrong).

Author: >As mentioned above, I would remove CREATE_MACHINE and DELETE_MACHINE and treat them as core capabilities that must be implemented

I find it is a valid point. However, I think there is no harm in having these are REQUIRED capabilities right? I see even the CSI following something similar. Refer https://github.com/container-storage-interface/spec/blob/master/csi.proto#L127-L134 and https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR122. And having the other methods as optional ones. 

However, I have updated comments & docs to indicate `CREATE_MACHINE` and `DELETE_MACHINE` as required capabilities. Also in the sample controller implementation which would be used to generate a blueprint code, I shall make sure that these capabilities are always returned.  I hope this is acceptable. 

> Also, I don't see a capability for LIST_MACHINES - is this intentional? I believe this call should be treated as non-core and therefore a capability for it should exist (but I may be wrong).

No you are right. `LIST_MACHINES` is non-core. That was a mistake. I have added it now.

            SHUTDOWN_MACHINE = 4;

            // GET_LIST_OF_VOLUMEIDS_FOR_EXISTING_PVS tells if the plugin
            // implements the GetListOfVolumeIDsForExistingPVs() interface
            GET_LIST_OF_VOLUMEIDS_FOR_EXISTING_PVS = 5;
        }
        Type type = 1;
    }
    oneof type {
        // RPC that the plugin supports.
        RPC rpc = 1;
    }
}
```

##### GetPluginCapabilities Errors

If the plugin is unable to complete the GetPluginCapabilities call successfully, it MUST return a non-ok gRPC code in the gRPC status.

#### `Probe`

A Plugin MUST implement this RPC call.
The primary utility of the Probe RPC is to verify that the plugin is in a healthy and ready state.
If an unhealthy state is reported, via a non-success response, a CMI-Client MAY take action with the intent to bring the plugin to a healthy state.
Such actions MAY include, but SHALL NOT be limited to, the following:

* Restarting the plugin container, or
* Notifying the plugin supervisor.

The Plugin MAY verify that it has the right configurations, devices, dependencies and drivers in order to run and return a success if the validation succeeds.
The CMI-Client MAY invoke this RPC at any time.

> A CMI-Client MAY invoke this call multiple times with the understanding that a plugin's implementation MAY NOT be trivial and there MAY be overhead incurred by such repeated calls.

Reviewer 3: `A CMI-Client MAY invoke this call multiple times with the understanding that a plugin's implementation MAY NOT be trivial and there MAY be overhead incurred by such repeated calls.`

Would it make sense to introduce a sleep for such repeated calls to the plugin container, similar behaviour was introduced in Etcd-backup-restore recently as a similar liveness check was running too often and was sub-optimal. wdyt ?

Author: The implementation of the `CMI-Client` would definitely make these calls after regular intervals (we are even thinking of retry details to be configurable). However, this is just a heads up to plugin implementor to be ready for simultaneous/multiple calls.

Author: I have added a section on the retry mechanism. https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR205-R211

The VM provider SHALL document guidance and known limitations regarding a particular Plugin's implementation of this RPC.
For example, the VM provider MAY document the maximum frequency at which its Probe implementation SHOULD be called.

```protobuf
message ProbeRequest {
    // Intentionally empty.
}

message ProbeResponse {
    // Readiness allows a plugin to report its initialization status back
    // to the CMI-Client. Initialization for some plugins MAY be time consuming
    // and it is important for a CMI-Client to distinguish between the following
    // cases:
    //
    // 1) The plugin is in an unhealthy state and MAY need restarting. In
    //    this case a gRPC error code SHALL be returned.
    // 2) The plugin is still initializing, but is otherwise perfectly
    //    healthy. In this case a successful response SHALL be returned
    //    with a readiness value of `false`. Calls to the plugin's

>     //    Controller and/or Node services MAY fail due to an incomplete

Reviewer 4: `Controller and/or Node services` should be replaced with `Machine` service.

Author: Done.

    //    initialization state.
    // 3) The plugin has finished initializing and is ready to service

>     //    calls to its Controller and/or Node services. A successful

Reviewer 4: `Controller and/or Node services` should be replaced with `Machine` service.

Author: Done.

    //    response is returned with a readiness value of `true`.
    //
    // This field is OPTIONAL. If not present, the caller SHALL assume
    // that the plugin is in a ready state and is accepting calls to its

>     // Controller and/or Node services (according to the plugin's reported

Reviewer 4: `Controller and/or Node services` should be replaced with `Machine` service.

Author: Done.

| Condition | gRPC Code | Description | Recovery Behavior |
|-----------|-----------|-------------|-------------------|
| Plugin not healthy | 9 FAILED_PRECONDITION | Indicates that the plugin is not in a healthy/ready state. | Caller SHOULD assume the plugin is not healthy and that future RPCs MAY fail because of this condition. |
| Missing required dependency | 9 FAILED_PRECONDITION | Indicates that the plugin is missing one or more required dependency. | Caller MUST assume the plugin is not healthy. |

### Machine Service RPC

#### `CreateMachine`

A Plugin MUST implement this RPC call if it has `CREATE_MACHINE` capability.
This RPC will be called by the CMI-Client to provision a new machine on behalf of a user.

- If a machine corresponding to the specified machine `name` already exists, and is compatible with the specified `ProviderSpec` in the `CreateMachineRequest`, the Plugin MUST reply `0 OK` with the corresponding `CreateMachineResponse`.
- The plugin can OPTIONALY make use of the secrets supplied in the `Secrets` map in the `CreateMachineRequest` to communicate with the provider.

> - The plugin must have an unique way to map a `machine` name to a `VM`. This could be implicitly provided by the provider by letting you set VM-names (or) could be explicitly specified by the plugin using appropriate tags to map the same.

Reviewer 4: s/an unique/a unique/

Author: Done.

- This operation MUST be idempotent.

- The `CreateMachineResponse` returned by this method is expected to return

>     - `MachineID` that uniquely identifys the VM at the provider. This is expected n the  matches with the node.Spec.ProviderID on the node object.

Reviewer 4: s/expected n the matches/expected to match/
(?)

Author: Done.

}
```

##### CreateMachine Errors

If the plugin is unable to complete the CreateMachine call successfully, it MUST return a non-ok gRPC code in the gRPC status.
If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
The CMI-Client MUST implement the specified error recovery behavior when it encounters the gRPC error code.

| gRPC Code | Condition | Description | Recovery Behavior | Auto Retry Required |
|-----------|-----------|-------------|-------------------|------------|
| 0 OK | Successful | The call was successful in creating/adopting a VM that matches supplied creation request. The `CreateMachineResponse` is returned with desired values |  | N |
| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `Name` and `ProviderSpec`. Make sure all parameters are in permitted range of values. Exact issue to be given in `.message` | Update providerSpec to fix issues. | N |

> | 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |

Reviewer 3: how will one come out of it if the supplied deadline is managed by cmi-client and not the provider as retry may never succeed if the provider takes more time then what is set as default timeout ? related to the comment I made in an earlier section.

Author: Okay, let's take an example of a machine-creation taking up to 20mins on a provider XYZ and machine creation timeout set to 5mins by `CMI-Client`. Does it make sense for the plugin to partially process creation request, and on deadline exceed continue the creation in the next creation request? So in this case, the machine creation may take 4-5 creation calls all of whom return non-0 error with retry except for the last call. 

Will this make sense to the plugin implementor?

Reviewer 5: @prashanth26 
Actually, this should work.

Reviewer 2: @prashanth26 This is exactly what makes sense IMO, and how we've already implemented it.

Author: @mkoynov @stoyanr - I had your situation in mind while drafting this. Will try to incorporate changes on MCM (CMI-Client) to accommodate this. I hope this is acceptable. Do raise concerns if you have any.

| 5 NOT_FOUND |  |  |  |  |
| 6 ALREADY_EXISTS | Already exists but desired parameters doesn't match | Parameters of the existing VM don't match the ProviderSpec | Create machine with a different name | N |
| 7 PERMISSION_DENIED | Insufficent permissions | The requestor doesn't have enough permissions to create an VM and it's required dependencies | Update requestor permissions to grant the same | N |
| 8 RESOURCE_EXHAUSTED | Resource limits have been reached | The requestor doesn't have enough resource limits to process this creation request | Enhance resource limits associated with the user/account to process this | N |
| 9 PRECONDITION_FAILED | VM is in inconsistent state | The VM is in a state that is invalid for this operation | Manual intervention might be needed to fix the state of the VM | N |

> | 10 ABORTED |  |  |  |  |

Reviewer 5: This error code should be handled as well as explained [here](https://github.com/prashanth26/machine-spec/blob/enhancement/updated-spec/spec.md#concurrency). Same for DeleteMachine.

Reviewer 2: Agree with @mkoynov , this error is possible and even likely, so it should be described how it's handled by the client. `Auto Retry Required` should be Yes.

| 11 OUT_OF_RANGE | Resources were out of range  | The requested number of CPUs, memory size, of FS size in ProviderSpec falls outside of the corresponding valid range | Update request paramaters to request valid resource requests | N |
| 12 UNIMPLEMENTED | Not implemented | Unimplemented indicates operation is not implemented or not supported/enabled in this service. | Retry with an alternate logic or implement this method at the plugin. Most methods by default are in this state | N |
| 13 INTERNAL | Major error | Means some invariants expected by underlying system has been broken. If you see one of these errors, something is very broken. | Needs manual intervension to fix this | N |
| 14 UNAVAILABLE | Not Available | Unavailable indicates the service is currently unavailable. | Retry operation after sometime | Y |

> | 15 DATALOSS |  |  |  |  |

Reviewer 3: empty row => not applicable ?

Author: Yes. It means not applicable. However, just wanted to keep the empty roll for anyone to fill up if it makes sense in this case.

| 0 OK | Successful | The call was successful in deleting a VM that matches supplied deletion request. |  | N |
| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `MachineID` and make sure that it is in the desired format and not a blank value. Exact issue to be given in `.message` | Update `MachineID` to fix issues. | N |
| 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |
| 5 NOT_FOUND |  |  |  |  |
| 6 ALREADY_EXISTS |  |  |  |  |
| 7 PERMISSION_DENIED | Insufficent permissions | The requestor doesn't have enough permissions to delete an VM and it's required dependencies | Update requestor permissions to grant the same | N |
| 8 RESOURCE_EXHAUSTED |  |  |  |  |
| 9 PRECONDITION_FAILED | VM is in inconsistent state | The VM is in a state that is invalid for this operation | Manual intervention might be needed to fix the state of the VM | N |

> | 10 ABORTED |  |  |  |  |

Reviewer 2: Also here, this error is possible and even likely, so it should be described how it's handled by the client. `Auto Retry Required` should be Yes.

        Running = 1;
        Stopped = 2;
    }
    Status MachineStatus = 2;
}
```

##### GetMachine Errors

If the plugin is unable to complete the GetMachine call successfully, it MUST return a non-ok gRPC code in the gRPC status.
If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.

| gRPC Code | Condition | Description | Recovery Behavior | Auto Retry Required |
|-----------|-----------|-------------|-------------------|------------|
| 0 OK | Successful | The call was successful in getting machine details for given `MachineID` |  | N |
| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `MachineID` and make sure that it is in the desired format and not a blank value. Exact issue to be given in `.message` | Update `MachineID` to fix issues. | N |
| 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |

> | 5 NOT_FOUND |  |  |  |  |

Reviewer 2: `NOT_FOUND` is entirely possible here, should be described.

| 6 ALREADY_EXISTS |  |  |  |  |
| 7 PERMISSION_DENIED | Insufficent permissions | The requestor doesn't have enough permissions to get details for the VM and it's required dependencies | Update requestor permissions to grant the same | N |
| 8 RESOURCE_EXHAUSTED |  |  |  |  |
| 9 PRECONDITION_FAILED | VM is in inconsistent state | The VM is in a state that is invalid for this operation | Manual intervention might be needed to fix the state of the VM | N |

> | 10 ABORTED |  |  |  |  |

Reviewer 2: Not sure about `ABORTED` for `GetMachines`. We would not return it in our implementation since it's harmless to query GMP machines in parallel. I think you should decide on one of the following 2 strategies and describe accordingly:
* `ABORTED` is valid only for operations that change the state of a machine.
* `ABORTED` is valid everywhere.

Author: I think option (1) is better. 

- ABORTED is valid only for operations that change the state of a machine.

| 0 OK | Successful | The termination request for the VM was accepted (or) the VM is already in terminated state. |  | N |
| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `MachineID` and make sure that it is in the desired format and not a blank value. Exact issue to be given in `.message` | Update `MachineID` to fix issues. | N |
| 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |
| 5 NOT_FOUND |  |  |  |  |
| 6 ALREADY_EXISTS |  |  |  |  |
| 7 PERMISSION_DENIED | Insufficent permissions | The requestor doesn't have enough permissions to shut down a VM and it's required dependencies | Update requestor permissions to grant the same | N |
| 8 RESOURCE_EXHAUSTED |  |  |  |  |
| 9 PRECONDITION_FAILED | VM is in inconsistent state | The VM is in a state that is invalid for this operation | Manual intervention might be needed to fix the state of the VM | N |

> | 10 ABORTED |  |  |  |  |

Reviewer 2: `ABORTED` is possible here, should be described.

| 11 OUT_OF_RANGE |  |  |  |  |  |
| 12 UNIMPLEMENTED | Not implemented | Unimplemented indicates operation is not implemented or not supported/enabled in this service. | Retry with an alternate logic or implement this method at the plugin. Most methods by default are in this state | N |
| 13 INTERNAL | Major error | Means some invariants expected by underlying system has been broken. If you see one of these errors, something is very broken. | Needs manual intervension to fix this | N |
| 14 UNAVAILABLE | Not Available | Unavailable indicates the service is currently unavailable. | Retry operation after sometime | Y |
| 15 DATALOSS |  |  |  |  |
| 16 UNAUTHENTICATED | Missing provider credentials | Request does not have valid authentication credentials for the operation | Fix the provider credentials | N |

The status `message` MUST contain a human readable description of error, if the status `code` is not `OK`.
This string MAY be surfaced by CMI-Client to end users.

#### `ListMachines`

A Plugin MUST implement this RPC call if it has `LIST_MACHINES` capability.
The Plugin SHALL return the information about all the machines associated with the `ProviderSpec`.

> The CMI-Client SHALL NOT expect a consistent "view" of all machines when paging through the machines list via multiple calls to `ListMachines`.

Reviewer 2: You mention "paging" here but don't actually describe a paging mechanism. Do you intend for the plugin to page responses, similarly to CSI spec? If yes, this should be described in more details.
I personally don't like pages since it makes things more complex for everybody, so if you think it can be avoided, by all means let's do without it. I am not sure however if that would work for all cases and providers, you might run into some limits related to the size of responses for very large sets.

Author: I second with you on this. I hadn't put too much thought into this in my initial draft. Now when I think of it, It's a complex solution. I shall remove this sentence and we shall stick to a simple response.

    // This field is REQUIRED.
    map<string, string> MachineList = 1;
}
```

##### ListMachines Errors

If the plugin is unable to complete the ListMachines call successfully, it MUST return a non-ok gRPC code in the gRPC status.
If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
The CMI-Client MUST implement the specified error recovery behavior when it encounters the gRPC error code.

| gRPC Code | Condition | Description | Recovery Behavior | Auto Retry Required |
|-----------|-----------|-------------|-------------------|------------|
| 0 OK | Successful | The call for listing all VMs associated with `ProviderSpec` was successful. |  | N |
| 1 CANCELED | Cancelled | Call was cancelled. Perform any pending clean-up tasks and return the call |  | N |
| 2 UNKNOWN | Something went wrong | Not enough information on what went wrong | Retry operation after sometime | Y |
| 3 INVALID_ARGUMENT | Re-check supplied parameters | Re-check the supplied `ProviderSpec` and make sure that all required fields are present in their desired value format. Exact issue to be given in `.message` | Update `ProviderSpec` to fix issues. | N |

> | 4 DEADLINE_EXCEEDED | Timeout | The call processing exceeded supplied deadline | Retry operation after sometime | Y |

Reviewer 3: Can timeout happen because of too many machines to be returned in the call ? Should we also recommend a relook at filters defined in provider spec to list machines ?

Author: I think it could be possible in a provider with 1000s or millions of VM managed and accessible by the `provider-credentials` given here. However, only a handful of them (~10) might be relevant to the supplied `provider-spec`.

 I shall add a sentence about filtering the instances in the above call.

Author: Added a line on it. - https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR735

| 8 RESOURCE_EXHAUSTED |  |  |  |  |
| 9 PRECONDITION_FAILED |  |  |  |  |
| 10 ABORTED |  |  |  |  |
| 11 OUT_OF_RANGE |  |  |  |  |  |
| 12 UNIMPLEMENTED | Not implemented | Unimplemented indicates operation is not implemented or not supported/enabled in this service. | Retry with an alternate logic or implement this method at the plugin. Most methods by default are in this state | N |
| 13 INTERNAL | Major error | Means some invariants expected by underlying system has been broken. If you see one of these errors, something is very broken. | Needs manual intervension to fix this | N |
| 14 UNAVAILABLE | Not Available | Unavailable indicates the service is currently unavailable. | Retry operation after sometime | Y |
| 15 DATALOSS |  |  |  |  |
| 16 UNAUTHENTICATED | Missing provider credentials | Request does not have valid authentication credentials for the operation | Fix the provider credentials | N |

The status `message` MUST contain a human readable description of error, if the status `code` is not `OK`.
This string MAY be surfaced by CMI-Client to end users.

> #### `GetListOfVolumeIDsForExistingPVs`

Reviewer 2: This naming is somewhat clumsy IMO. Perhaps you could use a shorter name, e.g. `GetVolumeIDs` or at least `GetVolumeIDsForExistingPVs`? Also, a sentence about why such a RPC call is useful, e.g. for what purpose it could be used by a client would be great, at least to me it's not entirely obvious.

A Plugin MUST implement this RPC call if it has `GET_LIST_OF_VOLUMEIDS_FOR_EXISTING_PVS` capability.
This RPC will be called by the CMI-Client to get the `VolumeIDs` for the list of `PersistantVolumes (PVs)` supplied.

- On succesful returnal of a list of `Volume-IDs` for all supplied `PVs`, the Plugin MUST reply `0 OK`.
- The `GetListOfVolumeIDsForExistingPVsResponse` is expected to return a repeated list of `strings` consisting of the `VolumeIDs` for `PVSpec` that could be extracted.

> - If for any `PV` the Plugin wasn't able to identify the `Volume-ID`, the plugin maybe chose to ignore it and return the `Volume-IDs` for the rest of the `PVs` for whom the `Volume-ID` was found.

Reviewer 4: s/maybe/may/

Author: Done.

    Proxy components transporting communication over IP networks SHALL be responsible for securing communications over such networks.
* Both the CMI-Client and Plugin SHOULD avoid accidental leakage of sensitive information (such as redacting such information from log files).

### Debugging

* Debugging and tracing are supported by external, CMI-independent additions and extensions to gRPC APIs, such as [OpenTracing](https://github.com/grpc-ecosystem/grpc-opentracing).

## Configuration and Operation

### General Configuration

* The `CMI_ENDPOINT` environment variable SHALL be supplied to the Plugin by the Plugin Supervisor.
* An operator SHALL configure the CMI-Client to connect to the Plugin via the listen address identified by `CMI_ENDPOINT` variable.
* With exception to sensitive data, Plugin configuration SHOULD be specified by environment variables, whenever possible, instead of by command line flags or bind-mounted/injected files.

> #### Plugin Bootstrap Example

Reviewer: The endpoint examples all seem to be unix socket endpoints. Can we have URL endpoints in the samples?

Author: Yes, that makes sense. Shall try to update the same.

Author: Updated.

* Supervisor -> Plugin: `CMI_ENDPOINT=unix:///path/to/unix/domain/socket.sock`.
* Operator -> CMI-Client: use plugin at endpoint `unix:///path/to/unix/domain/socket.sock`.
* CMI-Client: monitor `/path/to/unix/domain/socket.sock`.
* Plugin: read `CMI_ENDPOINT`, create UNIX socket at specified path, bind and listen.
* CMI-Client: observe that socket now exists, establish connection.
* CMI-Client: invoke `GetPluginCapabilities`.

#### Filesystem

> * Plugins SHALL NOT specify requirements that include or otherwise reference directories and/or files on the root filesystem of the CO.

Reviewer 5: What is CO?

> * Plugins SHALL NOT specify requirements that include or otherwise reference directories and/or files on the root filesystem of the CO.

Reviewer 4: s/CO/cmi client/

Author: Done.

Note: All UNIX endpoints SHALL end with `.sock`. See [gRPC Name Resolution](https://github.com/grpc/grpc/blob/master/doc/naming.md).

This variable is REQUIRED.

#### Operational Recommendations

The Plugin Supervisor expects that a Plugin SHALL act as a long-running service vs. an on-demand, CLI-driven process.

Supervised plugins MAY be isolated and/or resource-bounded.

##### Logging

* Plugins SHOULD generate log messages to ONLY standard output and/or standard error.
  * In this case the Plugin Supervisor SHALL assume responsibility for all log lifecycle management.
* Plugin implementations that deviate from the above recommendation SHALL clearly and unambiguously document the following:
  * Logging configuration flags and/or variables, including working sample configurations.
  * Default log destination(s) (where do the logs go if no configuration is specified?)
  * Log lifecycle management ownership and related guidance (size limits, rate limits, rolling, archiving, expunging, etc.) applicable to the logging mechanism embedded within the Plugin.
* Plugins SHOULD NOT write potentially sensitive data to logs (e.g. secrets).

##### Available Services

> * Plugin Packages MAY support all or a subset of CMI services; service combinations MAY be configurable at runtime by the Plugin Supervisor.

Reviewer: Can we describe what happens if some of the services or RPCs are not supported?

Author: Okay.

Reviewer 4: Not counting the identity service, we have only one service (machine) as of now. This may not be applicable to CMI. We can reword this as `Available RPCs`, but this is covered in detail in the preceding text anyway.

Author: Added it here - https://github.com/gardener/machine-spec/pull/11/files#diff-958e7270f96f5407d7d980f500805b1bR122

>   * A plugin MUST know the "mode" in which it is operating (e.g. node, controller, or both).

Reviewer 4: N/A for CMI

Author: Removed.

  * This specification does not dictate the mechanism by which mode of operation MUST be discovered, and instead places that burden upon the VM Provider.
* Misconfigured plugin software SHOULD fail-fast with an OS-appropriate error code.

##### Linux Capabilities

> * Plugin Supervisor SHALL guarantee that plugins will have `CAP_SYS_ADMIN` capability on Linux when running on Nodes.

Reviewer 4: This was applicable to CSI. However, it should not be required for CMI, right?

Author: Removed above sentence.

* Plugins SHOULD clearly document any additionally required capabilities and/or security context.

> ##### Namespaces

Reviewer 4: Again, this is applicable to CSI, and not CMI

Author: I think we can keep this for now.

Author: You are right this is not relevant. Hence removing this

> * A Plugin SHOULD NOT assume that it is in the same [Linux namespaces](https://en.wikipedia.org/wiki/Linux_namespaces) as the Plugin Supervisor.

Reviewer 4: Do we need to define `Plugin Supervisor` like it is done in CSI spec?

Author: You are right this is not relevant. Hence removing this

>   The CMI-Client MUST clearly document the [mount propagation](https://www.kernel.org/doc/Documentation/filesystems/sharedsubtree.txt) requirements for Node Plugins and the Plugin Supervisor SHALL satisfy the CO’s requirements.

Reviewer 4: This doesn't apply to CMI plugin. CSI needs this because its Node Plugin is responsible for un/staging volumes.

Author: You are right this is not relevant. Hence removing this

##### Cgroup Isolation

* A Plugin MAY be constrained by cgroups.

> * An operator or Plugin Supervisor MAY configure the devices cgroup subsystem to ensure that a Plugin MAY access requisite devices.

Reviewer 4: Is it relevant to CMI?

Author: I think we can keep this for now.

Author: You are right this is not relevant. Hence removing this

* A Plugin Supervisor MAY define resource limits for a Plugin.

##### Resource Requirements

* VM Providers SHOULD unambiguously document all of a Plugin’s resource requirements.

### Deploying
* **Recommended:** The CMI-Client and Plugin are typically expected to run as two containers inside a common `Pod`.

> * However, for the security reasons they could execute on seperate Pods provided they have an secure way to exchange data between them.

Reviewer 4: s/an secure/a secure/

Author: Done.
